### ISSUE TITLE
Destroying a socket does not release processes waiting on that socket

### ISSUE THREAD

**Issue by daniels220** [guillep/pharo-core#15975]
## Bug description
If I call one of the `#waitForData...` family and the other side closes the connection, the `readSemaphore` is signaled reasonably quickly and the method indicates the connection is closed. Similarly for `waitForSendDone`. However if another Smalltalk process calls `close`, and even `destroy`, on _this_ end of the connection, none of the Semaphores are ever signaled and any waiting process will be stuck until the garbage collector reclaims them or their timeouts expire.

## To Reproduce
Note, in all these examples, the other side of the connection is "unresponsive"—otherwise the operation would finish long before the timeout expires. This is just for convenience, there are plenty of cases where this might occur in normal operation—the listening socket of a webserver, or an open WebSocket or similar long-lived connection waiting for traffic that might not occur for many minutes.

### When listening
```smalltalk
Transcript clear.
sock := Socket newTCP
	        listenOn: 1234 backlogSize: 5;
	        yourself.
[
| waitTime |
Transcript crShow: 'Waiting for connection'.
waitTime := Time millisecondsToRun: [
	            sock
		            waitForConnectionFor: 5
		            ifTimedOut: [ Transcript crShow: 'Wait timed out' ] ].
Transcript crShow: 'Wait took ' , waitTime printString , 'ms' ] fork.
Transcript crShow: 'Forked proc started'.
2 seconds asDelay wait.
Transcript crShow: 'Closing socket'.
sock closeAndDestroy.
```
Times out after 5s rather than breaking loose after 2s. Note that the socket-closed case is actually entirely unhandled here and would be treated as a timeout even if the semaphore were signaled.

### When receiving
```smalltalk
Transcript clear.
listen := Socket newTCP
	          listenOn: 1234 backlogSize: 5;
	          yourself.
[
client := Socket newTCP
	          connectTo: (SocketAddress fromDottedString: '127.0.0.1')
	          port: 1234 ] fork.
server := listen waitForAcceptFor: 1.
[
| waitTime |
Transcript crShow: 'Waiting for data'.
waitTime := Time millisecondsToRun: [
	            client
		            waitForDataFor: 5
		            ifClosed: [ Transcript crShow: 'Connection closed' ]
		            ifTimedOut: [ Transcript crShow: 'Wait timed out' ] ].
Transcript crShow: 'Wait took ' , waitTime printString , 'ms' ] fork.
Transcript crShow: 'Forked proc started'.
2 seconds asDelay wait.
Transcript crShow: 'Closing socket'.
client closeAndDestroy.
0.5s seconds asDelay wait.
server closeAndDestroy.
listen closeAndDestroy.
```
In this case the `ifClosed:` branch is taken, but only after the 5-second timeout expires.

### When sending
Note, there isn't a terribly convenient way to send a lot of data with a custom timeout—I'd have to rewrite the outer loop of `sendData:`, so it's easier to just impose a shorter timeout externally—the point is that the process should be unblocked and report the closure much sooner.
```smalltalk
Transcript clear.
listen := Socket newTCP
	          listenOn: 1234 backlogSize: 5;
	          yourself.
[
client := Socket newTCP
	          connectTo: (SocketAddress fromDottedString: '127.0.0.1')
	          port: 1234 ] fork.
server := listen waitForAcceptFor: 1.
[
Transcript crShow: 'Beginning send'.
[
client sendData: (ByteArray new: 2 ** 24).
Transcript crShow: 'Send data completed' ]
	valueWithin: 5 seconds
	onTimeout: [ Transcript crShow: 'Wait timed out' ] ] fork.
Transcript crShow: 'Forked proc started'.
2 seconds asDelay wait.
Transcript crShow: 'Closing socket'.
client closeAndDestroy.
0.5s seconds asDelay wait.
server closeAndDestroy.
listen closeAndDestroy
```

## Expected behavior
When the socket is destroyed, all waiting processes should be unblocked and properly report that the socket has been closed out from under them.

Worth noting that many methods actually break in this circumstance, as they directly refer to the `socketHandle` without a nil-check, and/or call primitives that fail if the socket is in `InvalidSocket` state without checking that first.

## Version information
Any OS, any Pharo

## Expected development cost
It would be a start to send `signalAll` to each Semaphore before niling in it in `destroy`. There is possibly a race condition here as someone could have a reference to the Semaphore on the stack and be about to call `wait` when we call `signalAll`, though this seems very unlikely. Fixing the methods that break if the socket is closed out from under them is a bit more work, but not too difficult—basically any method that refers directly to the `socketHandle`, and also sends any message with `wait` in the name, is suspect. There's probably a way to bring up an actual browser on this, but an inspector will do:
```smalltalk
Socket methodDictionary select: [ :each |
	(each readsSlot: (Socket slotNamed: 'socketHandle')) and: [
		each ast sentMessages anySatisfy: [ :msg | msg beginsWith: #wait ] ] ]
```
In many cases the `wait...` method itself can perform the necessary check and raise an exception if the socket is closed, in which case the caller can remain unchanged. It's probably more in keeping with the rest of the API to add an `ifClosed:` branch to the `wait` methods that already have an `ifTimedOut:`, then implement a version throwing an exception on top of that. The API here is a bit messy and it would be nice to have a "style guide" of sorts for it (that's the wrong phrase—a set of rules/conventions the API follows, basically). I could probably help some with hashing out what questions need to be answered and documenting the way things are and what is decided, but this seems like something that needs input from a number of people to _make_ those decisions.

**Comment by tesonep** [comment]
Hi, thanks a lot for the catch. 
I think it is a good idea to handle it in the wait... and add a IfClosed: block as you said. 
I don't think there could be a race condition, but in that case nilling the variable should help to avoid most of them.
